src/main/java/cn/nukkit/block/BlockDoor.java

```diff
@@ -153,27 +158,101 @@ private void onNormalUpdate() {
                 level.setBlock(this, Block.get(AIR), false);
             }
 
+            /* Doesn't work with redstone
             boolean downIsOpen = down.getBooleanValue(OPEN);
             if (downIsOpen != isOpen()) {
                 setOpen(downIsOpen);
                 level.setBlock(this, this, false, true);
-            }
+            }*/
             return;
         }
-        
+
         if (down.getId() == AIR) {
             level.useBreakOn(this, getToolType() == ItemTool.TYPE_PICKAXE? Item.get(ItemID.DIAMOND_PICKAXE) : null);
         }
     }
 
+    @PowerNukkitDifference(info = "Checking if the door was opened/closed manually.")
     private void onRedstoneUpdate() {
-        boolean isPowered = level.isBlockPowered(this.getLocation());
-        if (isOpen() != isPowered) {
-            level.getServer().getPluginManager().callEvent(new BlockRedstoneEvent(this, isOpen() ? 15 : 0, isOpen() ? 0 : 15));
+        if ((this.isOpen() != this.isGettingPower()) && !this.getManualOverride()) {
+            if (this.isOpen() != this.isGettingPower()) {
+                level.getServer().getPluginManager().callEvent(new BlockRedstoneEvent(this, this.isOpen() ? 15 : 0, this.isOpen() ? 0 : 15));
+
+                this.setOpen(null, this.isGettingPower());
+            }
+        } else if (this.getManualOverride() && (this.isGettingPower() == this.isOpen())) {
+            this.setManualOverride(false);
+        }
+    }
+
+    @PowerNukkitOnly
+    @Since("1.4.0.0-PN")
+    public void setManualOverride(boolean val) {
+        Location down;
+        Location up;
+        if (this.isTop()) {
+            down = down().getLocation().clone();
+            up = getLocation().clone();
+        } else {
+            down = getLocation().clone();
+            up = up().getLocation().clone();
+        }
+
+        if (val) {
+            manualOverrides.add(up);
+            manualOverrides.add(down);
+        } else {
+            manualOverrides.remove(up);
+            manualOverrides.remove(down);
+        }
+    }
+
+    @PowerNukkitOnly
+    @Since("1.4.0.0-PN")
+    public boolean getManualOverride() {
+        Location down;
+        Location up;
+        if (this.isTop()) {
+            down = down().getLocation().clone();
+            up = getLocation().clone();
+        } else {
+            down = getLocation().clone();
+            up = up().getLocation().clone();
+        }
+
+        return manualOverrides.contains(up) || manualOverrides.contains(down);
+    }
+
+    @Override
+    @PowerNukkitOnly
+    @Since("1.4.0.0-PN")
+    public boolean isGettingPower() {
+        Location down;
+        Location up;
+        if (this.isTop()) {
+            down = down().getLocation().clone();
```

Comment:
Why you need to deep copy the location of the block? wouldn't just using the reference be enough for this use case?

Reply:
yes would be, I'll edit it